airbrakes/data_handling/data_processor.py

```diff
@@ -245,7 +268,33 @@ def _calculate_speeds(self, a_x: list[float], a_y: list[float], a_z: list[float]
         velocities_z: np.array = previous_vel_z + np.cumsum(np.array(a_z) * time_diff)
 
         # Store the last calculated velocity vectors
-        self._previous_velocity = (velocities_x[-1], velocities_y[-1], velocities_z[-1])
+        self._previous_velocity_from_acceleration = (velocities_x[-1], velocities_y[-1], velocities_z[-1])
 
         # All the speeds gotten as the magnitude of the velocity vector at each point
-        return np.sqrt(velocities_x**2 + velocities_y**2 + velocities_z**2)
+        return np.sqrt(velocities_x ** 2 + velocities_y ** 2 + velocities_z ** 2)
+
+    def _calculate_speeds_from_altitude(self, altitude_list: list[float]) -> npt.NDArray[np.float64]:
+        """
+        Calculates the speed of the rocket based on the altitude data points. The pressure altitude is only
+        actually updated at a rate of 50hz, so we have to check that it's a "new" data point before calculating speed.
+        """
+        # Accounts for start up not having a last data point
+        if self._last_data_point is None:
+            return np.array([0.0] * len(self._data_points))
+
+        # Because the pressure altitude is only updated at 50Hz, we have to check if the data point is new, we do this
+        # by assuming that we will almost never get the same altitude reading twice in a row, so if we get the unique
+        # altitudes, that should be each altitude reading. In theory, we don't actually need to do this as the math will
+        # work out, but it's better to be safe than sorry and I like how this is more explict/intentional.
+        altitude_list.append(self._last_data_point.estPressureAlt)
+        unique_altitudes = np.unique(altitude_list)
```

Comment:
do not modify the original passed in list (pressure_altitudes). Always make a new copy. Also, why is the last data point of the pressure alt added to the end instead of the beginning?